### PR TITLE
Prevent `AtomicBool::fetch_and` and `fetch_or` from CAS loop on x86

### PR DESCRIPTION
**Pull request by AngelicosPhosphoros**
Currently, this operations generate Compare-And-Swap loop on x86 and x86_64.

Example:

```asm
    movzx        eax, byte ptr [rdi]
.LBB1_1:
    mov          ecx, eax
    and          cl, sil
    lock cmpxchg byte ptr [rdi], cl
    jne          .LBB1_1
    test         al, al
    setne        al
    ret
```

If we exploit the fact that AtomicBool can be either `0` and `1`, we can avoid this loop, and, in cases when argument is known, any branching at all.

```asm
    test    esi, esi
    je      .LBB1_1
    mfence
    movzx   eax, byte ptr [rdi]
    jmp     .LBB1_2
.LBB1_1:
    xor     eax, eax
    xchg    byte ptr [rdi], al
.LBB1_2:
    test    al, al
    setne   al
    ret
```

We still need to use `swap` and `fetch_add` instead of `store` and `load` because we need to get old value and apply provided memory ordering.

Related discussion: https://internals.rust-lang.org/t/should-we-change-fetch-logic-op-for-atomicbool-to-exploit-knowledge-about-it-contents/20098

### CODE REVIEW

**Comment by AngelicosPhosphoros** [comment]
Also, there is a [suggestion](https://internals.rust-lang.org/t/should-we-change-fetch-logic-op-for-atomicbool-to-exploit-knowledge-about-it-contents/20098/9) to use `compare_exchange` here to remove branching at all.
It also always has SeqCst semantics on x86.

```rust
fn fetch_and(a: &AtomicBool, val: bool, order: Ordering)->bool {
    let (Ok(v)|Err(v)) = a.compare_exchange(true, val, SeqCst);
    v
}

fn fetch_or(a: &AtomicBool, val: bool, order: Ordering)->bool {
    let (Ok(v)|Err(v)) = a.compare_exchange(false, val, SeqCst);
    v
}
```

File: library/core/src/sync/atomic.rs [hunk 1]
```diff
@@ -904,8 +904,22 @@ impl AtomicBool {
     #[cfg(target_has_atomic = "8")]
     #[cfg_attr(miri, track_caller)] // even without panics, this helps for Miri backtraces
     pub fn fetch_and(&self, val: bool, order: Ordering) -> bool {
-        // SAFETY: data races are prevented by atomic intrinsics.
-        unsafe { atomic_and(self.v.get(), val as u8, order) != 0 }
+        if cfg!(any(target_arch = "x86_64", target_arch = "x86")) {
```
Comment (Nilstrieb) [inline library/core/src/sync/atomic.rs:907]: Are there other architectures that also generate a needless CAS loop? Might be worth leaving a comment with all the arches where this _doesn't_ matter, so people in the future know that their arch wasn't forgotten.

Comment (AngelicosPhosphoros) [reply library/core/src/sync/atomic.rs:907]: There are some, I added list of those that doesn't need this.

**Comment by AngelicosPhosphoros** [comment]
r? Nilstrieb
Well, you have already commented...

**Comment by Nilstrieb** [comment]
r? Amanieu

**Comment by Amanieu** [comment]
I like the basic idea, however I would like the code to be cleaned up a bit. Currently we already have a similar hack to deal with RISC-V (`EMULATE_ATOMIC_BOOL`). It would be confusing to have 2 different ways of "emulating" AtomicBool operations.

One possible way forward would be to:
- rename `EMULATE_ATOMIC_BOOL` to `EMULATE_ATOMIC_BOOL_RISCV`.
- add `EMULATE_ATOMIC_BOOL_X86` for this, with a comment explaining it.

I'm not too sure about the naming though, so I'm open to other suggestions.

**Comment by AngelicosPhosphoros** [comment]
I would update PR on the weekends.

**Comment by Amanieu** [comment]
ping @AngelicosPhosphoros

**Comment by Dylan-DPC** [comment]
@AngelicosPhosphoros any updates on this? thanks

**Comment by Dylan-DPC** [comment]
Closing this as inactive. Feel free to reöpen this pr or create a new pr if you get the time to work on this. Thanks

**Pull request closed without merging**